File: src/positioning/libqgeoshape.cpp
#include <QGeoCoordinate>
#include <QGeoRectangle>
#include <QGeoShape>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qgeoshape.h>
#include "libqgeoshape.h"
#include "libqgeoshape.hxx"

QGeoShape* QGeoShape_new() {
    return new QGeoShape();
}

QGeoShape* QGeoShape_new2(const QGeoShape* other) {
    return new QGeoShape(*other);
}

int QGeoShape_Type(const QGeoShape* self) {
    return static_cast<int>(self->type());
}

bool QGeoShape_IsValid(const QGeoShape* self) {
    return self->isValid();
}

bool QGeoShape_IsEmpty(const QGeoShape* self) {
    return self->isEmpty();
}

bool QGeoShape_Contains(const QGeoShape* self, const QGeoCoordinate* coordinate) {
    return self->contains(*coordinate);
}

QGeoRectangle* QGeoShape_BoundingGeoRectangle(const QGeoShape* self) {
    return new QGeoRectangle(self->boundingGeoRectangle());
}

QGeoCoordinate* QGeoShape_Center(const QGeoShape* self) {
    return new QGeoCoordinate(self->center());
}

void QGeoShape_OperatorAssign(QGeoShape* self, const QGeoShape* other) {
    self->operator=(*other);
}

libqt_string QGeoShape_ToString(const QGeoShape* self) {
    QString _ret = self->toString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QGeoShape_Delete(QGeoShape* self) {
    delete self;
}

File: include/positioning/libqgeoshape.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QGeoCoordinate = @import("libqt6").QGeoCoordinate;
const QGeoRectangle = @import("libqt6").QGeoRectangle;
const qgeoshape_enums = enums;
const std = @import("std");

/// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html)
pub const QGeoShape = extern struct {
    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.QGeoShape,

    pub const _is_QGeoShape = {};

    /// New constructs a new QGeoShape object.
    ///
    pub fn New() QGeoShape {
        return .{ .ptr = qtc.QGeoShape_new() };
    }

    /// New2 constructs a new QGeoShape object.
    ///
    /// ## Parameter(s):
    ///
    /// ` other: QGeoShape `
    ///
    pub fn New2(other: anytype) QGeoShape {
        comptime _ = @TypeOf(other)._is_QGeoShape;
        return .{ .ptr = qtc.QGeoShape_new2(@ptrCast(other.ptr)) };
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#type)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    /// ## Returns:
    ///
    /// ` qgeoshape_enums.ShapeType `
    ///
    pub fn Type(self: QGeoShape) i32 {
        return qtc.QGeoShape_Type(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#isValid)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    pub fn IsValid(self: QGeoShape) bool {
        return qtc.QGeoShape_IsValid(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#isEmpty)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    pub fn IsEmpty(self: QGeoShape) bool {
        return qtc.QGeoShape_IsEmpty(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#contains)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    /// ` coordinate: QGeoCoordinate `
    ///
    pub fn Contains(self: QGeoShape, coordinate: anytype) bool {
        comptime _ = @TypeOf(coordinate)._is_QGeoCoordinate;
        return qtc.QGeoShape_Contains(@ptrCast(self.ptr), @ptrCast(coordinate.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#boundingGeoRectangle)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    pub fn BoundingGeoRectangle(self: QGeoShape) QGeoRectangle {
        return .{ .ptr = qtc.QGeoShape_BoundingGeoRectangle(@ptrCast(self.ptr)) };
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#center)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    pub fn Center(self: QGeoShape) QGeoCoordinate {
        return .{ .ptr = qtc.QGeoShape_Center(@ptrCast(self.ptr)) };
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#operator-eq)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    /// ` other: QGeoShape `
    ///
    pub fn OperatorAssign(self: QGeoShape, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QGeoShape;
        qtc.QGeoShape_OperatorAssign(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#toString)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoShape `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn ToString(self: QGeoShape, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.QGeoShape_ToString(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qgeoshape.ToString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `Delete` instead
    ///
    pub const QDelete = Delete;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#dtor.QGeoShape)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ## Parameter:
    ///
    /// ` self: QGeoShape `
    ///
    pub fn Delete(self: QGeoShape) void {
        qtc.QGeoShape_Delete(@ptrCast(self.ptr));
    }
};

/// ### [Upstream resources](https://doc.qt.io/qt-6/qgeoshape.html#public-types)
pub const enums = struct {
    pub const ShapeType = enum(i32) {
        pub const UnknownType: i32 = 0;
        pub const RectangleType: i32 = 1;
        pub const CircleType: i32 = 2;
        pub const PathType: i32 = 3;
        pub const PolygonType: i32 = 4;
    };
};
